compiler/aarch64/codegen/OMRMemoryReference.cpp

```diff
@@ -850,19 +850,68 @@ uint8_t *OMR::ARM64::MemoryReference::generateBinaryEncoding(TR::Instruction *cu
          }
       else
          {
-         // loadaddrEvaluator() uses addimmx in generateTrgMemInstruction
+         // loadaddrEvaluator() uses addimmx in generateTrg1MemInstruction
          TR_ASSERT(index == NULL, "MemoryReference with unexpected indexed form");
 
-         base->setRegisterFieldRN(wcursor);
-
          if (constantIsUnsignedImm12(displacement))
             {
             *wcursor |= (displacement & 0xfff) << 10; /* imm12 */
+            base->setRegisterFieldRN(wcursor);
             cursor += ARM64_INSTRUCTION_LENGTH;
             }
          else
             {
-            TR_ASSERT_FATAL(false, "Offset is too large for specified instruction.");
```

Comment:
I can see this as a temporary solution to get something to work, but I'm a bit hesitant to call this the long-term solution.  This code appears to make an arbitrary spilling decision without considering the actual availability of the register being spilled.  Given the number of registers available on this architecture, there's probably a good chance that in many cases the register is already free in which case the spill is unneeded.

This is a problem faced by the other code generators too (realizing that a register needs to be used after register allocation/assignment has occurred) and each solves the problem differently depending on the context.  In this particular situation, one way to solve it might be that when this memory reference is register assigned and it knows that there is the possibility that a register might be needed during binary encoding, it could ask the machine for the current register state and record one of the free registers just in case it might be needed.  The binary encoding would use that free register without a spill, or arbitrarily spill some register if there were none available.

This PR is probably OK as an initial contribution, but I would like an issue opened to track the work to implement something that makes more careful use of registers.

Reply:
I don't like this solution very much, either.  I followed what p and arm versions do.

(p) https://github.com/eclipse/omr/blob/d77a173dd7275268225917b0be3bf85ed5ef6b1f/compiler/p/codegen/OMRMemoryReference.cpp#L1406-L1429
(arm) https://github.com/eclipse/omr/blob/d77a173dd7275268225917b0be3bf85ed5ef6b1f/compiler/arm/codegen/OMRMemoryReference.cpp#L1216-L1224

Reply:
Can you open an issue as I requested above so that we can move this along please?

Reply:
I opened Issue #4862 for this.